### PR TITLE
Misalign storage device

### PR DESCRIPTION
**Pull request by Orycterope**

### CODE REVIEW

**Comment by Orycterope** [comment]
I was thinking about storing the underlying libuser::error in `BlockDeviceError`. This means making `BlockDeviceError` generic, and adding an associated `T::Error` on every `BlockDevice`. 

Do you have any argument for/against this ?

**Comment by roblabla** [comment]
I’d be for having an associated error type. The current enum is really lacking - in NSPs, there are multiple reasons for read failure: either the underlying file op failed, or the hashes don’t match. This information is currently lost at the blockdevice layer

File: src/storage_device.rs [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+            // truncate the buffer to only the interesting part so we're sure we don't spill.
+            let buf = &mut buf[first_part_len..(first_part_len + middle_part_len)];
+
+            let buf_misalignment = &mut buf[0] as *mut u8 as usize % align_of::<BD::Block>();
+
+            if middle_part_len > 0 {
+                if buf_misalignment == 0 {
+                    // read everything in one go
+                    // cast the buffer as an array of bytes
+                    let blocks = unsafe {
```
Comment (roblabla) [inline src/storage_device.rs]: Use plain here. There's 0 reason to use unsafe when BD::Block implements plain.

File: src/block_device.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    type Block = Block;
+
+    /// Seeks to the appropriate position, and reads block by block.
+    fn read(&mut self, blocks: &mut [Block], index: BlockIndex) -> Result<(), BlockDeviceError> {
+        use std::io::{Read, Seek};
+
+        self.seek(std::io::SeekFrom::Start(index.0 * size_of::<Self::Block>() as u64)).ok()
+            .and_then(|_| {
+                let blocks_as_slice = unsafe {
+                    core::slice::from_raw_parts_mut(blocks as *mut _ as *mut u8, blocks.len() * size_of::<Self::Block>())
```
Comment (roblabla) [inline src/block_device.rs]: Safety comment. Should say something along the lines of "Block has no padding".

File: src/block_device.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            })
+    }
+
+    /// Seeks to the appropriate position, and writes block by block.
+    fn write(&mut self, blocks: &[Block], index: BlockIndex) -> Result<(), BlockDeviceError> {
+        use std::io::{Seek, Write};
+
+        self.seek(std::io::SeekFrom::Start(index.0 * size_of::<Self::Block>() as u64)).ok()
+            .and_then(|_| {
+                let blocks_as_slice = unsafe {
```
Comment (roblabla) [inline src/block_device.rs]: ditto

**Review by marysaka — APPROVED** [review]

**Review by roblabla — APPROVED** [review]

**Pull request merged**